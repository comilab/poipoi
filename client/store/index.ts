import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
