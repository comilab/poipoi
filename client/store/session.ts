import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
