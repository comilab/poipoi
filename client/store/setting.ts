import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
