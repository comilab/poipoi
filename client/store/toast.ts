import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
