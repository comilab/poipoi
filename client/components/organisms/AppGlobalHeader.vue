<template lang="pug">
header.bg-white.text-sm.lg_sticky.top-0.lg_shadow.z-40.px-2.xl_px-8.py-2
  .flex.flex-wrap.items-center
    .block.lg_hidden
      AppButton(
        to="/",
        color="blue",
        :icon="faHome"
      )
    .hidden.lg_block.ml-2
      h1.font-bold
        AppButton(
          to="/",
          color="blue",
          size="lg"
        ) {{ store.setting.data.siteTitle }}
    form.flex-auto.flex.items-center.ml-4.mr-8(
      @submit.prevent="search",
      role="search"
    )
      AppInput.flex-auto.w-20.sm_w-auto(
        type="search",
        v-model="input.keyword",
        placeholder="検索"
      )
      AppButton.ml-2(
        type="submit",
        color="white",
        :icon="faSearch",
        pill
      )
    .block.md_hidden.relative
      AppButton(
        color="white",
        :icon="faBars",
        pill,
        @click="showMenu = !showMenu"
      )
    nav.w-full.md_w-auto.overflow-hidden.md_overflow-visible(:class="{'h-0 md_h-auto': !showMenu}")
      ul.grid.gap-2.md_gap-0.divide-y.md_divide-y-0.md_flex.md_justify-end.pt-2.md_pt-0
        template(v-if="store.session.isLogin")
          li.md_ml-2
            AppButton.w-full.md_w-auto(
              color="white",
              :icon="faPencilAlt",
              to="/posts/new",
              pill
            ) 投稿
          li.relative.pt-2.md_pt-0.md_ml-2
            AppButton.w-full.md_w-auto(
              to="/notifications",
              color="white",
              :icon="faBell",
              pill
            ) お知らせ
          li.pt-2.md_pt-0.md_ml-2
            AppButton.w-full.md_w-auto(
              to="/settings",
              color="white",
              :icon="faCog",
              pill
            ) 設定
          li.pt-2.md_pt-0.md_ml-2
            AppButton.w-full.md_w-auto(
              color="white",
              :icon="faSignOutAlt",
              pill,
              :indeterminate="logoutSending",
              :disabled="logoutSending",
              @click="logout"
            ) ログアウト
        template(v-else)
          li.pt-2.md_pt-0.md_ml-2
            AppButton.w-full.md_w-auto(
              to="/login",
              color="white",
              :icon="faSignInAlt",
              pill
            ) ログイン
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs, useContext, watch } from 'nuxt-composition-api'
import { faBars, faHome, faSearch, faPencilAlt, faBell, faCog, faSignInAlt, faSignOutAlt } from '@fortawesome/free-solid-svg-icons'

import { useStore } from '~/store'
import AppButton from '~/components/atoms/AppButton.vue'
import AppInput from '~/components/molecules/AppInput.vue'

export default defineComponent({
  components: {
    AppInput,
    AppButton
  },
  setup () {
    const { route, redirect } = useContext()

    const store = useStore()

    const localState = reactive({
      input: {
        keyword: route.value.query.keyword || ''
      },
      showMenu: false,
      logoutSending: false
    })

    const search = () => {
      redirect(200, '/', {
        ...localState.input
      })
    }

    const logout = async () => {
      localState.logoutSending = true
      await store.session.destroy()
      store.toast.push({
        type: 'success',
        message: 'ログアウトしました'
      })
      redirect(200, '/')
    }

    watch(
      () => route.value.query.keyword,
      (keyword) => {
        localState.input.keyword = keyword
      }
    )

    return {
      store,
      ...toRefs(localState),
      search,
      logout,
      faBars,
      faHome,
      faSearch,
      faPencilAlt,
      faBell,
      faCog,
      faSignInAlt,
      faSignOutAlt
    }
  }
})
</script>

<style lang="postcss" scoped>
.top-half {
  top: 50%;
}

nav {
  & ul {
    & li {
      & > * {
        @apply justify-start;

        &.nuxt-link-active {
          @apply bg-gray-300;
        }

        @screen md {
          @apply justify-center;
        }
      }
    }
  }
}
</style>
