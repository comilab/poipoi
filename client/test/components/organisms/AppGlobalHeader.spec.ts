import { fireEvent, render, waitFor, within } from '@testing-library/vue'
import * as NuxtCompositionApi from 'nuxt-composition-api'
import { ref } from '@vue/composition-api'

import AppGlobalHeader from '~/components/organisms/AppGlobalHeader.vue'
import * as store from '~/store'
import settingFactory from '~/test/factories/setting'
import userFactory from '~/test/factories/user'

describe('components/organisms/AppGlobalHeader', () => {
  let options: any
  let contextMock: any

  beforeEach(() => {
    options = {
      stubs: {
        FontAwesomeIcon: true,
        NuxtLink: true
      }
    }

    store.store.setting.setData(settingFactory.build())
    jest.spyOn(store, 'useStore').mockReturnValue(store.store)

    contextMock = {
      redirect: jest.fn(),
      route: ref<any>({
        query: {}
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useContext').mockReturnValue(contextMock)
  })

  test('renders correctly', () => {
    store.store.setting.setData(settingFactory.build({
      siteTitle: 'site-title'
    }))
    const { container } = render(AppGlobalHeader, options)

    expect(container).toMatchSnapshot()
  })

  describe('検索', () => {
    test('送信するとリダイレクトする', async () => {
      const { getByRole } = render(AppGlobalHeader, options)
      await fireEvent.update(getByRole('searchbox'), 'foo')
      await fireEvent.submit(getByRole('search'))

      expect(contextMock.redirect).toBeCalledWith(200, '/', { keyword: 'foo' })
    })

    test('route.query.keywordの値を入力欄にセット', async () => {
      const { getByRole } = render(AppGlobalHeader, options)

      expect(getByRole('searchbox')).toHaveValue('')

      contextMock.route.value = {
        query: { keyword: 'foo' }
      }

      await waitFor(() => {
        expect(getByRole('searchbox')).toHaveValue('foo')
      })
    })
  })

  describe('ログアウト', () => {
    beforeEach(() => {
      jest.spyOn(store.store.session, 'destroy').mockResolvedValue()
      store.store.session.setUser(userFactory.build())
    })

    test('クリックするとログアウト処理を実行', async () => {
      const { getByRole } = render(AppGlobalHeader, options)
      await fireEvent.click(getByRole('button', { name: 'ログアウト' }))

      expect(store.store.session.destroy).toBeCalled()
    })

    test('処理中はdisabled, indeterminate', async () => {
      const { getByRole } = render(AppGlobalHeader, options)
      await fireEvent.click(getByRole('button', { name: 'ログアウト' }))

      expect(getByRole('button', { name: 'ログアウト' })).toHaveAttribute('disabled')
      expect(within(getByRole('button', { name: 'ログアウト' })).getByRole('status')).toBeInTheDocument()
    })

    test('処理後にリダイレクト', async () => {
      const { getByRole } = render(AppGlobalHeader, options)
      await fireEvent.click(getByRole('button', { name: 'ログアウト' }))

      await waitFor(() => {
        expect(contextMock.redirect).toBeCalledWith(200, '/')
      })
    })
  })
})
