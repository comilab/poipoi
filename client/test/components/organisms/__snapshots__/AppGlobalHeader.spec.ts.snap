// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppGlobalHeader renders correctly 1`] = `
<div>
  <header
    class="bg-white text-sm lg_sticky top-0 lg_shadow z-40 px-2 xl_px-8 py-2"
  >
    <div
      class="flex flex-wrap items-center"
    >
      <div
        class="block lg_hidden"
      >
        <nuxtlink-stub
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 square-md rounded"
          data-testid="button"
          to="/"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          />
        </nuxtlink-stub>
      </div>
      <div
        class="hidden lg_block ml-2"
      >
        <h1
          class="font-bold"
        >
          <nuxtlink-stub
            class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 text-xl px-4 py-2 rounded"
            data-testid="button"
            to="/"
          >
            <!---->
            <div
              class=""
              data-testid="text"
            >
              site-title
            </div>
          </nuxtlink-stub>
        </h1>
      </div>
      <form
        class="flex-auto flex items-center ml-4 mr-8"
        role="search"
      >
        <div
          class="flex flex-col flex-auto w-20 sm_w-auto untouched pristine"
          data-testid="input-null"
          placeholder="検索"
        >
          <!---->
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-0"
            placeholder="検索"
            type="search"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <button
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none ml-2 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded-full"
          data-testid="button"
          type="submit"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          />
        </button>
      </form>
      <div
        class="block md_hidden relative"
      >
        <button
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 square-md rounded-full"
          data-testid="button"
          type="button"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          />
        </button>
      </div>
      <nav
        class="w-full md_w-auto overflow-hidden md_overflow-visible h-0 md_h-auto"
      >
        <ul
          class="grid gap-2 md_gap-0 divide-y md_divide-y-0 md_flex md_justify-end pt-2 md_pt-0"
        >
          <li
            class="pt-2 md_pt-0 md_ml-2"
          >
            <nuxtlink-stub
              class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none w-full md_w-auto bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded-full"
              data-testid="button"
              to="/login"
            >
              <div>
                <fontawesomeicon-stub
                  class="fa-fw mr-2"
                  data-testid="icon"
                  icon="[object Object]"
                />
              </div>
              <div
                class=""
                data-testid="text"
              >
                ログイン
              </div>
            </nuxtlink-stub>
          </li>
        </ul>
      </nav>
    </div>
  </header>
</div>
`;
